# **Hospital Readmission Rates - ETL Process**

## Objectives

* Extract dataset from Kaggle and import into notebook
* Push data through ETL pipeline to clean data and use Numpy or Pandas where required

## Inputs

* Dataset from raw data file, imported from Kaggle

## Outputs

* Clean Data, exported to "Clean Data" file



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\naqas\\OneDrive\\Documents\\Coding\\CI_Projects\\hospital_readmission_rates_analysis\\hospital_readmission_rates_analysis\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\naqas\\OneDrive\\Documents\\Coding\\CI_Projects\\hospital_readmission_rates_analysis\\hospital_readmission_rates_analysis'

# Import required packages

Import required packages for ETL

In [5]:
import pandas as pd
import numpy as np

---

# Import and clean dataset

## _Removal of Redundant Columns, Null values and Duplicates_

The dataset must be imported and then transferred to a dataframe.

In [6]:
raw_hospital_data_df = pd.read_csv("data_files/RawData/diabetic_data.csv")
raw_hospital_data_df

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
101761,443847548,100162476,AfricanAmerican,Male,[70-80),?,1,3,7,3,...,No,Down,No,No,No,No,No,Ch,Yes,>30
101762,443847782,74694222,AfricanAmerican,Female,[80-90),?,1,4,5,5,...,No,Steady,No,No,No,No,No,No,Yes,NO
101763,443854148,41088789,Caucasian,Male,[70-80),?,1,1,7,1,...,No,Down,No,No,No,No,No,Ch,Yes,NO
101764,443857166,31693671,Caucasian,Female,[80-90),?,2,3,7,10,...,No,Up,No,No,No,No,No,Ch,Yes,NO


There are a lot of columns included in this dataset so it is important to obtain all the information about the dataset before continuing the cleaning.

In [7]:
raw_hospital_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

Now, we will run a null count.

In [8]:
raw_hospital_data_df.isnull().sum()

encounter_id                    0
patient_nbr                     0
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                          0
diag_3                          0
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

From the initial output of the dataframe we can see that the dataframe references null values as "?". Pandas does not recognise this and so this value has to be replaced in the dataframe as, when the null count has been run above, it is showing columns to have 0 null values even though we know there are some. To do this we will replace "?" with "NaN".

In [9]:
raw_hospital_data_df = raw_hospital_data_df.replace("?",np.nan)
raw_hospital_data_df

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),NaN,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),NaN,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),NaN,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),NaN,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),NaN,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
101761,443847548,100162476,AfricanAmerican,Male,[70-80),NaN,1,3,7,3,...,No,Down,No,No,No,No,No,Ch,Yes,>30
101762,443847782,74694222,AfricanAmerican,Female,[80-90),NaN,1,4,5,5,...,No,Steady,No,No,No,No,No,No,Yes,NO
101763,443854148,41088789,Caucasian,Male,[70-80),NaN,1,1,7,1,...,No,Down,No,No,No,No,No,Ch,Yes,NO
101764,443857166,31693671,Caucasian,Female,[80-90),NaN,2,3,7,10,...,No,Up,No,No,No,No,No,Ch,Yes,NO


Next, we will do a null count.

In [10]:
raw_hospital_data_df.isnull().sum()

encounter_id                    0
patient_nbr                     0
race                         2273
gender                          0
age                             0
weight                      98569
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                  40256
medical_specialty           49949
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                         21
diag_2                        358
diag_3                       1423
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

We can see now that the null count values have changed as we have implemented the replacement. This will make the data cleaning more accurate.

We will be dropping the following columns and the reasons why:

* Medical Specialty - Not required for analysis 
* Weight - Too many null values, redundant column
* Payer code - Not required for analysis
* Encounter ID - Not required for analysis
* Patient Number - Not required for analysis
* Admission Type ID - ID conversion not possible as ID codes not available, redundant column
* Admission Source ID - ID conversion not possible as ID codes not available, redundant column
* Number of Diagnoses - ID conversion not possible as ID codes not available, redundant column


In [11]:
raw_hospital_data_df = raw_hospital_data_df.drop(["medical_specialty","payer_code", "weight", "encounter_id", "patient_nbr","admission_type_id", "discharge_disposition_id", "admission_source_id","number_diagnoses"], axis = 1)
raw_hospital_data_df

,race,gender,age,time_in_hospital,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,Caucasian,Female,[0-10),1,41,0,1,0,0,0,...,No,No,No,No,No,No,No,No,No,NO
1,Caucasian,Female,[10-20),3,59,0,18,0,0,0,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,AfricanAmerican,Female,[20-30),2,11,5,13,2,0,1,...,No,No,No,No,No,No,No,No,Yes,NO
3,Caucasian,Male,[30-40),2,44,1,16,0,0,0,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,Caucasian,Male,[40-50),1,51,0,8,0,0,0,...,No,Steady,No,No,No,No,No,Ch,Yes,NO
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
101761,AfricanAmerican,Male,[70-80),3,51,0,16,0,0,0,...,No,Down,No,No,No,No,No,Ch,Yes,>30
101762,AfricanAmerican,Female,[80-90),5,33,3,18,0,0,1,...,No,Steady,No,No,No,No,No,No,Yes,NO
101763,Caucasian,Male,[70-80),1,53,0,9,1,0,0,...,No,Down,No,No,No,No,No,Ch,Yes,NO
101764,Caucasian,Female,[80-90),10,45,2,21,0,0,1,...,No,Up,No,No,No,No,No,Ch,Yes,NO


Next, there are null values in the "Race" column. As this will be analysed, the null records must also be removed.

In [12]:
raw_hospital_data_df = raw_hospital_data_df.dropna(subset =["race"])
raw_hospital_data_df.isnull().sum()

race                            0
gender                          0
age                             0
time_in_hospital                0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                         19
diag_2                        336
diag_3                       1349
max_glu_serum               94203
A1Cresult                   82897
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide                   0
glipizide                       0
glyburide                       0
tolbutamide                     0
pioglitazone                    0
rosiglitazone                   0
acarbose                        0
miglitol                        0
troglitazone                    0
tolazamide    

Under the diagnoses columns, the null values only exist if there was no concurrent diagnosis. However, there should be no nulls in the primary diagnosis column (diag_1). These must be removed.

In [13]:
raw_hospital_data_df = raw_hospital_data_df.dropna(subset =["diag_1"])
raw_hospital_data_df.isnull().sum()

race                            0
gender                          0
age                             0
time_in_hospital                0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                        335
diag_3                       1348
max_glu_serum               94185
A1Cresult                   82881
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide                   0
glipizide                       0
glyburide                       0
tolbutamide                     0
pioglitazone                    0
rosiglitazone                   0
acarbose                        0
miglitol                        0
troglitazone                    0
tolazamide    

In the remaining 4 columns, nulls can be replaced with "none" or "0". For the diagnosis columns this is because it is not necessary to have a second or third diagnosis. A patient is able to just have a single diagnosis. With the "Max glucose serum" and "A1C result" columns, this is because in these records it is considered that the measurement was not taken. This potentially has an impact on patient readmittance and so the nulls must be replaced with a value so that visualisation can be done.

In [14]:
raw_hospital_data_df["diag_2"] = raw_hospital_data_df["diag_2"].fillna(0)
raw_hospital_data_df["diag_3"] = raw_hospital_data_df["diag_3"].fillna(0)
raw_hospital_data_df.isnull().sum()

race                            0
gender                          0
age                             0
time_in_hospital                0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                          0
diag_3                          0
max_glu_serum               94185
A1Cresult                   82881
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide                   0
glipizide                       0
glyburide                       0
tolbutamide                     0
pioglitazone                    0
rosiglitazone                   0
acarbose                        0
miglitol                        0
troglitazone                    0
tolazamide    

As I will be binarising the "Max glucose serum" and "A1C result" columns, I will be replacing the nulls with "0".

In [15]:
raw_hospital_data_df["max_glu_serum"] = raw_hospital_data_df["max_glu_serum"].fillna(0)
raw_hospital_data_df["A1Cresult"] = raw_hospital_data_df["A1Cresult"].fillna(0)
raw_hospital_data_df.isnull().sum()

race                        0
gender                      0
age                         0
time_in_hospital            0
num_lab_procedures          0
num_procedures              0
num_medications             0
number_outpatient           0
number_emergency            0
number_inpatient            0
diag_1                      0
diag_2                      0
diag_3                      0
max_glu_serum               0
A1Cresult                   0
metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol                    0
troglitazone                0
tolazamide                  0
examide                     0
citoglipton                 0
insulin                     0
glyburide-

Now that all the null values have been removed, it is essential to check for any duplicate records as this could cause skewing in data and affect the performance of any machine learning models.

In [16]:
raw_hospital_data_df[raw_hospital_data_df.duplicated(keep=False)]

,race,gender,age,time_in_hospital,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
67130,Caucasian,Female,[50-60),3,7,2,17,1,0,0,...,No,Down,No,No,No,No,No,Ch,Yes,NO
86916,Caucasian,Female,[50-60),3,7,2,17,1,0,0,...,No,Down,No,No,No,No,No,Ch,Yes,NO


There is one duplicate row in this data. This will be removed.

In [17]:
raw_hospital_data_df = raw_hospital_data_df.drop_duplicates()
raw_hospital_data_df.duplicated().sum()

0

Finally, gender has a value of "unknown/invalid". These rows must be removed as the values are too vague and will affect the data analysis.

In [18]:
raw_hospital_data_df= raw_hospital_data_df[raw_hospital_data_df["gender"] != "Unknown/Invalid"]
raw_hospital_data_df["gender"].unique()

array(['Female', 'Male'], dtype=object)

In [19]:
raw_hospital_data_df["diag_1"].isnull().sum()

0

## _Data Conversion_

The three diagnoses columns are categorised by ICD9 codes for illnesses. ICD9 is the "International Statistical Classification of Diseases and Health Related Problems". This is medical classification used in epidemiology, health management and clinical diagnosis. It is a classification system maintained by the World Health Organisation(WHO).

In this dataset, illnesses are classified by their ICD9 codes however, this is not useful information for data analysis. To fix this, we will convert the codes into the following categories:

* Circulatory
* Respiratory
* Digestive
* Diabetes
* Injury
* Musculoskeletal
* Genitourinary
* Neoplasms
* Endocrine
* Other

To complete this, a mapping function must be written to map the first 3 numbers of each ICD9 code to the medical categories to which they are related. Firstly, as there are codes that begin with string values of "E" or "V" then they need to be dealt with first. After which, the method defines "code" as an integer where the first 3 numbers of the integer are important. The "if,elif,else" block runs through all the categories and codes to change each entry. Finally, the for loop allows us to run through each column running the mapping function.

In [20]:
def map_category(code):
    code = str(code).split(".")[0]
    
    if code.startswith("E") or code.startswith("V"):
        return "Other"
    
    try:
        code = int(code[:3])
    except ValueError:
        return None
    
    if 1 <= code <= 139:
        return "Other"
    elif 140 <= code <= 239:
        return "Neoplasms"
    elif 240 <= code <= 249:
        return "Endocrine"
    elif code == 250:
        return "Diabetes"
    elif 251 <= code <= 279:
        return "Endocrine"
    elif 280 <= code <= 389:
        return "Other"
    elif 390 <= code <= 459:
        return "Circulatory"
    elif 460 <= code <= 519:
        return "Respiratory"
    elif 520 <= code <= 579:
        return "Digestive"
    elif 580 <= code <= 629:
        return "Genitourinary"
    elif 630 <= code <= 679:
        return "Other"
    elif 680 <= code <= 709:
        return "Other"
    elif 710 <= code <= 739:
        return "Musculoskeletal"
    elif 740 <= code <= 759:
        return "Other"
    elif code in [780, 781, 782, 783, 784, 789]:
        return "Other"
    elif code == 785:
        return "Circulatory"
    elif code == 786:
        return "Respiratory"
    elif code == 787:
        return "Digestive"
    elif code == 788:
        return "Genitourinary"
    elif 790 <= code <= 799:
        return "Other"
    elif 800 <= code <= 999:
        return "Injury"
    else:
        return None

cols = ["diag_1", "diag_2", "diag_3"]

for col in cols:
    raw_hospital_data_df[col] = raw_hospital_data_df[col].apply(map_category)

Once the function is run we can check if it works by calling back one of the columns from the dataframe and seeing the unique values.

In [21]:
raw_hospital_data_df["diag_1"].unique()

array(['Diabetes', 'Endocrine', 'Other', 'Neoplasms', 'Circulatory',
       'Respiratory', 'Injury', 'Genitourinary', 'Musculoskeletal',
       'Digestive'], dtype=object)

We can see that the codes have been converted into categorical values.

In [22]:
diag_1_null_count = raw_hospital_data_df["diag_1"].isnull().sum()
diag_2_null_count = raw_hospital_data_df["diag_2"].isnull().sum()
diag_3_null_count = raw_hospital_data_df["diag_3"].isnull().sum()

print(f"The null values in the column for Diagnosis 1 are: {diag_1_null_count}")
print(f"The null values in the column for Diagnosis 2 are: {diag_2_null_count}")
print(f"The null values in the column for Diagnosis 3 are: {diag_3_null_count}")

The null values in the column for Diagnosis 1 are: 0
The null values in the column for Diagnosis 2 are: 335
The null values in the column for Diagnosis 3 are: 1348


After making this change, we can see that there are null values in the "Diagnosis 2" and "Diagnosis 3" columns. This was dealt with earlier in the ETL process however, when converting the codes to categorical values, the function has converted non-code values to nulls. We will replace the nulls with a value of "None" as it is possible for patients to have had no second or third diagnosis.

In [23]:
raw_hospital_data_df["diag_2"] = raw_hospital_data_df["diag_2"].fillna("None")
raw_hospital_data_df["diag_3"] = raw_hospital_data_df["diag_3"].replace(np.nan,"None")

print(f"The null values in the column for Diagnosis 2 are: {diag_2_null_count}")
print(f"The null values in the column for Diagnosis 3 are: {diag_3_null_count}")

The null values in the column for Diagnosis 2 are: 335
The null values in the column for Diagnosis 3 are: 1348


It is still showing the columns to contain nulls but this is due to the value becoming "None". A simple null count of the whole dataframe will show that there are no null values in the required columns.

In [24]:
raw_hospital_data_df.isnull().sum()

race                        0
gender                      0
age                         0
time_in_hospital            0
num_lab_procedures          0
num_procedures              0
num_medications             0
number_outpatient           0
number_emergency            0
number_inpatient            0
diag_1                      0
diag_2                      0
diag_3                      0
max_glu_serum               0
A1Cresult                   0
metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol                    0
troglitazone                0
tolazamide                  0
examide                     0
citoglipton                 0
insulin                     0
glyburide-

In [25]:
print(f"The unique values in the Diagnosis 2 column are:{raw_hospital_data_df["diag_2"].unique()}")
print(f"The unique values in the Diagnosis 3 column are:{raw_hospital_data_df["diag_3"].unique()}")

The unique values in the Diagnosis 2 column are:['None' 'Diabetes' 'Neoplasms' 'Circulatory' 'Respiratory' 'Other'
 'Injury' 'Musculoskeletal' 'Genitourinary' 'Endocrine' 'Digestive']
The unique values in the Diagnosis 3 column are:['None' 'Endocrine' 'Other' 'Circulatory' 'Diabetes' 'Respiratory'
 'Injury' 'Neoplasms' 'Genitourinary' 'Digestive' 'Musculoskeletal']


Also, by running a "unique()" function to check the unique values in both the Diagnosis 2 and Diagnosis 3 columns, we can see that "None" is a value in both columns and there are no nulls.

---

# Export Cleaned Dataset

Now that the dataset has been cleaned, it is time to export it to prepare to use it for visualisations.

In [27]:
raw_hospital_data_df.to_csv("data_files/CleanData/clean_hospital_data.csv", index=False)

---

# Process Summary

This was a large dataset and had almost 50 different columns that required cleaning. The process chosen was as follows:

* A removal of redundant columns, null values and duplicates was the first important step:
    * The dataset had "?" in place where there should be null values. This required converting the "?" to a null value that "Pandas" would recognise.
    * Then before removing all the null values the redundant columns were to be removed. These columns were classified as redundant if:
        1) They did not have enough information to be used
        2) The information they had would have little value to visualisation or analysis
    * The columns marked as redundant were:
        * Medical Specialty - this was not relevant to the analysis and would provide no value.
        * Weight - this column had almost 90% null values and so had to be removed.
        * Payer code - this was not relevant to the analysis and would provide no value.
        * Encounter ID - this was just an ID column and would not be relevant to the analysis.
        * Patient Number - this was just an ID column and would not be relevant to the analysis.
        * Admission Type ID - this was a coded column but there was no information for the code conversion so the column provided no value.
        * Admission Source ID - this was a coded column but there was no information for the code conversion so the column provided no value.
        * Number of Diagnoses - this was a coded column but there was no information for the code conversion so the column provided no value.
    * The "Race" column contained null values and so these were removed from the dataset.
    * Next, the Diagnoses columns (Diag_1, Diag_2 & Diag_3) all contained some null values. A null value in diag_1 made no sense as there had to be an initial primary diagnosis. Any record with a null in this column was removed. For Diag_2 & Diag_3 however, it was not necessary to have a secondary or tertiary diagnosis. This meant, the nulls had to be converted to a "0" value. This conversion was done.
    * The columns "Max Glucose Serum" and "A1C result" had a lot of null values but this was important information. Part of the analysis to be conducted is to see the impact on patients when having these tests done against not having them performed. As a result of this it was again important to convert the nulls to a "0" value that could be used in the analysis as a comparison.
    * Next, a check was done for any duplicated records. Only one appeared. This was removed.
    *  In the "Gender" column there was an "Unknown/Invalid" option for the records. This had to be removed as the value was too vague for analysis.
* After this part of the cleaning, a data conversion process was required for the diagnoses columns as they contained integers with no clear value. All the values were ICD9 codes. ICD9 is the "International Statistical Classification of Diseases and Health Related Problems". This is medical classification used in epidemiology, health management and clinical diagnosis. It is a classification system maintained by the World Health Organisation(WHO). 
* It meant that the codes needed to be converted to categorical values. It was done with the following process:
    1) Firstly, it was determined what values would map to what categorical value (illnesses)
    2) Once this was determined a function was created with the following rules:
        * The function was named with an object called "code". "Code" would represent the ICD9 codes in the dataset.
        * The next part of the function would covert the code to a string and then split the code into parts at the decimal point.
        * Using a "startswith()" function, if the code started with "E" or "V" then the value would be returned as "Other" and therefore, those codes would be classified in the "Other" category.
        * Next, a "Try and Except" block is written to convert the code to an integer and only take the first three numbers in the code e.g. if the code is "1033", then it would be taken as "103". This is because the ICD9 codes are only 3 numbers long. The "Except" section of the code is input so that if there are any "Value" errors then the code block will be exited and the record marked as "None" as it would not match the requirements.
        * After this, an "if,elif,else" block was written to map the ICD9 codes to the categories they were related to. It seemed the best way to do it as the "Other" category absorbed a large variety of codes and was not just one single range.
        * Next, a temporary list called "cols" was defined containing the three diagnosis columns.
        * A "for" loop was created in which the "map_category()" function was applied via the "apply()" function. The for loop ran through each column one-by-one, running the mapping function each time. Use of the "apply()" function helped with running through each record in each column.
    * After the function was run, the unique values for the first column were checked to confirm the change to categoric values.
    * A null sum check was also completed to check for any null values. Upon doing this, there were 335 and 1348 in the "Diag_2" and "Diag_3" columns respectively.
    * This was initally dealt with earlier in the ETL however, the data conversion process and mapping function written later created null values again.
    * These null values were changed to "None" and then a null sum check was completed again.
    * The null sum still showed the same number of null values in each column, however, when doing the check against the whole dataframe, there were no nulls.
    * This showed that pandas was recognising "None" as a null value. When the unique values for both columns was checked, this confirmed that "None" was a value and there were no null values in both columns.
* Once this was all completed, the file was exported as a "Clean" data file, ready for use in the EDA.